homeassistant/components/bluetooth_tracker/device_tracker.py

```diff
@@ -65,103 +65,131 @@ def discover_devices(device_id: int) -> List[Tuple[str, str]]:
+async def get_tracking_devices(hass: HomeAssistantType) -> Tuple[Set[str], Set[str]]:
     """
     Load all known devices.
 
     We just need the devices so set consider_home and home range to 0
     """
     yaml_path: str = hass.config.path(YAML_DEVICES)
-    devices_to_track: Set[str] = set()
-    devices_to_not_track: Set[str] = set()
-
-    for device in run_coroutine_threadsafe(
-        async_load_config(yaml_path, hass, 0), hass.loop
-    ).result():
-        # Check if device is a valid bluetooth device
-        if not is_bluetooth_device(device):
-            continue
-
-        normalized_mac: str = device.mac[3:]
-        if device.track:
-            devices_to_track.add(normalized_mac)
-        else:
-            devices_to_not_track.add(normalized_mac)
+
+    devices = await async_load_config(yaml_path, hass, 0)
+    bluetooth_devices = [device for device in devices if is_bluetooth_device(device)]
+
+    devices_to_track: Set[str] = {
+        device.mac[3:] for device in bluetooth_devices if device.track
+    }
+    devices_to_not_track: Set[str] = {
+        device.mac[3:] for device in bluetooth_devices if not device.track
+    }
 
     return devices_to_track, devices_to_not_track
 
 
-def setup_scanner(hass: HomeAssistantType, config: dict, see, discovery_info=None):
+def lookup_name(mac: str) -> Optional[str]:
+    """Lookup a Bluetooth device name."""
+    _LOGGER.debug("Scanning %s", mac)
+    return bluetooth.lookup_name(mac, timeout=5)
+
+
+async def async_setup_scanner(
+    hass: HomeAssistantType, config: dict, async_see, discovery_info=None
+):
     """Set up the Bluetooth Scanner."""
     device_id: int = config.get(CONF_DEVICE_ID)
-    devices_to_track, devices_to_not_track = get_tracking_devices(hass)
+    interval = config.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL)
+    request_rssi = config.get(CONF_REQUEST_RSSI, False)
+    update_bluetooth_lock = asyncio.Lock()
 
     # If track new devices is true discover new devices on startup.
     track_new: bool = config.get(CONF_TRACK_NEW, DEFAULT_TRACK_NEW)
-    _LOGGER.debug("Tracking new devices = %s", track_new)
+    _LOGGER.debug("Tracking new devices is set to %s", track_new)
+
+    devices_to_track, devices_to_not_track = await get_tracking_devices(hass)
 
     if not devices_to_track and not track_new:
         _LOGGER.debug("No Bluetooth devices to track and not tracking new devices")
 
     if track_new:
-        for mac, device_name in discover_devices(device_id):
+        devices = await hass.async_add_executor_job(discover_devices, device_id)
+        for mac, device_name in devices:
             if mac not in devices_to_track and mac not in devices_to_not_track:
                 devices_to_track.add(mac)
-                see_device(see, mac, device_name)
 
-    interval = config.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL)
+            if mac in devices_to_track:
+                await see_device(hass, async_see, mac, device_name)
```

Comment:
We could optimize this by using `asyncio.wait` to await all these tasks concurrently.
```py
tasks = []
for mac, device_name in devices:
    ...
    tasks.append(see_device(hass, async_see, mac, device_name))

if tasks:
    await asyncio.wait(tasks)
```

Reply:
Nice 👏